pkg/resource/managed_reconciler.go

```diff
@@ -224,13 +238,15 @@ type mrManaged struct {
 	ManagedConnectionPublisher
 	ManagedEstablisher
 	ManagedFinalizer
+	ManagedInitializer
 }
 
 func defaultMRManaged(m manager.Manager) mrManaged {
 	return mrManaged{
 		ManagedConnectionPublisher: NewAPISecretPublisher(m.GetClient(), m.GetScheme()),
 		ManagedEstablisher:         NewAPIManagedFinalizerAdder(m.GetClient()),
 		ManagedFinalizer:           NewAPIManagedFinalizerRemover(m.GetClient()),
+		ManagedInitializer:         NewManagedNameAsExternalName(m.GetClient()),
```

Comment:
Could we merge the initializer and establisher interfaces into one? I prefer the name Initializer. We could either support chaining establishers, or just have one establisher that sets the external name annotation and adds a finalizer in order to reduce the amount of `Update` calls we must make.

Reply:
That was my first thought as well - move `Establish` up before the first `Observe` call and implement a new establisher.

At the time, I believe we put the `Establish` call in `!observation.ResourceExists` because we didn't want to set the finalizer of a resource that we didn't create in the first place and we have to do the naming assignment before the first `Observe` call. That's why I didn't go with that idea.

However, I can see that our thinking around managing existing resources started to change (see https://github.com/crossplaneio/crossplane-runtime/issues/27). Seems to me, our direction for existing resource will not have much to do with the finalizers after all. So, I'm happy to merge them into one.

Naming still has to be done before the `Observe` call, so, I'll refactor the existing establisher to be an initializer and remove the `ManagedEstablisher` completely.